### PR TITLE
SI-8611 Avoid accidental patmat unification with refinement types

### PR DESCRIPTION
**Pull request by retronym**
In the enclosed test, t8611a.scala, the patterns `O.{A, B}` were
incorrect treated as equivelent by the combination of
`uniqueTpForTree` and `Const.uniqueTpForTree`.

`uniqueTpForTree` used `Type#narrow` to try to create a distinct
type for each new pattern tree it encountered. However, narrowing a
`RefinedType` does not create a distinct type as we are used to
when narrowing, e.g. a class type.

```
// Type
def narrow: Type =
  if (phase.erasedTypes) this
  else {
    val cowner = commonOwner(this)
    refinedType(this :: Nil, cowner, EmptyScope, cowner.pos).narrow
  }

// CompoundType
override def narrow: Type = typeSymbol.thisType
```

This commit first tries to use `narrow`, and then detects if the
result shares a type symbol with the input. If so, it compensates
by creating another refinement type over the result and narrows that.

I've included a unit test to show the subtleties of narrowing refinements.

Review by @adriaanm. I'm hoping you can find a less ugly way to do this!

### CODE REVIEW

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by adriaanm** [comment]
I reported the IDE failure (deprecationWarningsDoNotAccumulate_1001595 ) to the team. Seeing it in multiple PRs.

File: src/compiler/scala/tools/nsc/transform/patmat/Logic.scala [hunk 1]
```diff
@@ -538,23 +538,32 @@ trait ScalaLogic extends Interface with Logic with TreeAndTypeAnalysis {
...
       // hashconsing trees (modulo value-equality)
-      private[TreesAndTypesDomain] def uniqueTpForTree(t: Tree): Type =
+      private[TreesAndTypesDomain] def uniqueTpForTree(t: Tree): Type = {
         // a new type for every unstable symbol -- only stable value are uniqued
         // technically, an unreachable value may change between cases
         // thus, the failure of a case that matches on a mutable value does not exclude the next case succeeding
         // (and thuuuuus, the latter case must be considered reachable)
-        if (!t.symbol.isStable) t.tpe.narrow
+        def narrowUniquely(tp: Type): Type = {
+          val tp1 = tp.narrow
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/Logic.scala]: Maybe we should avoid narrow and directly create a fresh abstract type symbol every time that's upper-bounded by tp? I didn't think this through fully, but maybe this need to use the "try a regular X or fallback to the X of something weird that's guaranteed to satisfy my requirements" pattern shows that X was not the right solution to begin with, and we should use "Y that satisfies my requirements".

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/Logic.scala]: what say you, @retronym?

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/patmat/Logic.scala]: Yeah, makes sense. I'll try that.

**Comment by retronym** [comment]
Re-review, please, by @adriaanm

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**